File: c-fisher/models/gift.js
import {HTTP} from '../utils/http.js'
class GiftModel extends HTTP{

  constructor(){
    super()
    this.url_prefix = 'gift'
  }

  /**
   * 当撤消自己的赠送清单时，需要更新属性gifts显示数据，防止多次发送请求，使用以下方式该更新数据
   */ 
  update(gifts,id) {
    let trades = gifts.trades
    for(let index in trades){
      let bid = trades[index].book.id
      // 找到要撤消的书籍,让后过滤掉
      if(bid === id){
        trades.splice(index,1)
        gifts.total -= 1
        break
      }
    }
  }


  /**
   * 撤销自己的赠送清单
   * bookId:图书唯一标识
   */
  cancel(bookId){
    return this.request({
      url:`${this.url_prefix}/cancel/${bookId}`,
      method:'DELETE'
    })
  }

  /**
   * 查询自己的赠送清单
   */
  getGiftsOfSelf(){
    return this.request({
      url:this.url_prefix+'/self'
    })
  }

  /**
   * 添加至赠送清单
   * bookId:图书唯一标识
   */
  save_of_gift(bookId){
    return this.request({
      url:`${this.url_prefix}/save/${bookId}`
    })
  }

  /**
   * 获取最近上传图书
   * start:分页查询起始索引
   */
  getRecentlyUploaded(start){
    return this.request({
      url:`${this.url_prefix}/recent?start=${start}`
    })
  }

  /**
   * 获取书籍拥有者信息
   * gid:礼物唯一标识
   */
  getOccupant(gid){
    return this.request({
      url:`${this.url_prefix}/occupant/${gid}`
    })
  }

}

export {
  GiftModel
}

File: c-fisher-api/app/api/v1/gift.py
from flask import g, jsonify, request, current_app
from lin.exception import Success

from app.libs.token_auth import auth
from lin.redprint import Redprint

from app.models.gift import Gift
from app.models.wish import Wish
from app.view_models.trade import MyTrade

gift_api = Redprint('gift')

@gift_api.route('/save/<int:bid>', methods=['GET'])
@auth.login_required
def save_of_gift(bid):
    uid = g.user.id
    gift = Gift(bid=bid, uid=uid)
    gift.save_of_gift()
    return Success()

@gift_api.route('/cancel/<int:bid>', methods=['DELETE'])
@auth.login_required
def cancel(bid):
    uid = g.user.id
    gift = Gift(bid=bid, uid=uid)
    gift.cancel()
    return Success()

@gift_api.route('/recent', methods=['GET'])
def get_recently_upload():
    start = request.args['start']
    pre_page_count = current_app.config['PRE_PAGE_COUNT']
    recently_upload = Gift.paged_query(start, pre_page_count, soft=True, launched=False)
    books = [r_u.book for r_u in recently_upload]
    total = Gift.get_total(soft=True, launched=False)
    result = {
        'books': books,
        'total': total
    }
    return jsonify(result)

@gift_api.route('/self', methods=['GET'])
@auth.login_required
def get_gifts_of_self():
    uid = g.user.id
    gifts = Gift.get_gifts_of_self(uid)
    trades_count_list = Wish.get_wished_count()
    my_trade = MyTrade(gifts, trades_count_list)
    return jsonify(my_trade)

@gift_api.route('/occupant/<int:gid>', methods=['GET'])
@auth.login_required
def get_occupant(gid):
    # 获取书籍拥有者信息
    consumer = Gift.query.get(gid).consumer
    t = {
        'nickname': consumer.wx_name,
        'beans': consumer.beans,
        'receive_or_send': str(consumer.receive_counter) + '/' + str(consumer.send_counter)
    }
    return jsonify(t), 200
